Implement generate command

andreasgerstmayr opened the pull request:
This debug utility reads a CR and outputs the manifests which a running instance of this operator would apply to the cluster.

Resolves: #166
Signed-off-by: Andreas Gerstmayr <agerstmayr@redhat.com>

andreasgerstmayr commented on cmd/generate/main.go:
```diff
@@ -0,0 +1,136 @@
+	println("usage: go run ./cmd/generate/main.go -cr path/to/microservices/cr.yaml")
+}
+
+func main() {
```
Alternatively we could also split up the main tempo-operator binary in multiple sub commands using a library like spf13/cobra.

pavolloffay replied:
I would prefer that, it will be easier to use it (not additional build is required)

rubenvp8510 replied:
I agree with this, having a sub commands will be better.

rubenvp8510 replied:
@andreasgerstmayr are you going to do this change? WDYT?

andreasgerstmayr replied:
@rubenvp8510 yes, it's almost done. I just need to update a few places where the tempo-operator executable is started (in the manifests and Makefile).
I'll also add an e2e test and send a PR tomorrow.

andreasgerstmayr replied:
Done:
```
$ go run main.go                      
Usage:
  tempo-operator [command]

Available Commands:
  completion  Generate the autocompletion script for the specified shell
  generate    Generate YAML manifests from a Tempo CR
  help        Help about any command
  start       Start the Tempo operator

Flags:
      --config string   The controller will load its initial configuration from this file. Omit this flag to use the default configuration values. Command-line flags override configuration from this file.
  -h, --help            help for tempo-operator

Use "tempo-operator [command] --help" for more information about a command.
```

:)

iblancasa reviewed (approved):
For the Jaeger Operator, there was an E2E test for this feature. Would it make sense to add something similar here?

frzifus commented on cmd/generate/main.go:
```diff
@@ -0,0 +1,136 @@
+	defer file.Close()
+
+	spec := v1alpha1.Microservices{}
+	decoder := yaml.NewYAMLOrJSONDecoder(file, 8192)
```
`8192` is a memory limiation? ^^

andreasgerstmayr replied:
Docs say `bufferSize determines how far into the stream the decoder will look to figure out whether this is a JSON stream (has whitespace followed by an open brace).`
Imho it shouldn't have to look very far to determine if the input file is a JSON or YAML.

rubenvp8510 replied:
Can we put this number in a constant? and a comment indicating what it is?

andreasgerstmayr replied:
Done.

frzifus commented on cmd/generate/main.go:
```diff
@@ -0,0 +1,136 @@
+}
+
+func usage() {
+	println("usage: go run ./cmd/generate/main.go -cr path/to/microservices/cr.yaml")
```
Is there a reason to use the builtin `println` here?

```
 Current implementations provide several built-in functions useful during bootstrapping. These functions are documented for completeness but are not guaranteed to stay in the language. They do not return a result. 
```
https://go.dev/ref/spec#Bootstrapping

andreasgerstmayr replied:
Ah I didn't realize it's not an official command. I've updated the PR to use sub commands, and removed this function.

andreasgerstmayr commented on apis/tempo/v1alpha1/microservices_webhook.go:
```diff
@@ -33,13 +33,19 @@ var microserviceslog = logf.Log.WithName("microservices-resource")
 
 //+kubebuilder:webhook:path=/mutate-tempo-grafana-com-v1alpha1-microservices,mutating=true,failurePolicy=fail,sideEffects=None,groups=tempo.grafana.com,resources=microservices,verbs=create;update,versions=v1alpha1,name=mmicroservices.kb.io,admissionReviewVersions=v1
 
+func NewDefaulter(defaultImages ImagesSpec) *defaulter {
```
The godoc check is disabled by default in golang-lint: https://github.com/golangci/golangci-lint/issues/456

I've added a comment and exported the struct now.

andreasgerstmayr commented:
I've moved the command to a subcommand now, and added an e2e test.
Unfortunately the e2e test currently fails on Kubernetes 1.21 because it cannot parse the `availableReplicas` field in `status` of `v1.StatefulSet`.

Is there a better way to drop the status field other than marshalling the object, then unmarshalling to a map, removing the field, and marshalling again (to YAML)? :confused:

iblancasa commented:
> Unfortunately the e2e test currently fails on Kubernetes 1.21 because it cannot parse the availableReplicas field in status of v1.StatefulSet.

So, instead of matching the `StatefulSet`, how about matching the `Pods` managed by that `StatefulSet`?

andreasgerstmayr commented on tests/e2e/generate/cr.yaml:
```diff
@@ -0,0 +1,8 @@
+apiVersion: tempo.grafana.com/v1alpha1
```
This is the input CR for the generate command in this e2e test. Basically like https://github.com/jaegertracing/jaeger-operator/blob/main/tests/e2e/generate/generate/jaeger-template.yaml.template, but for tempo.

iblancasa replied:
Oh, yes. Sorry. I didn't see the call in the previous file correctly.

rubenvp8510 commented on cmd/generate/main_test.go:
```diff
@@ -0,0 +1,65 @@
+	var buf bytes.Buffer
+	err := toYAML(scheme, []client.Object{&cm}, &buf)
+	require.NoError(t, err)
+	require.Equal(t, `---
```
woth to use `YAMLEq` instead?

andreasgerstmayr replied:
Good idea, I've updated the PR.

rubenvp8510 reviewed (approved):
Minor comments LGTM

andreasgerstmayr commented:
> > Unfortunately the e2e test currently fails on Kubernetes 1.21 because it cannot parse the availableReplicas field in status of v1.StatefulSet.
> 
> So, instead of matching the `StatefulSet`, how about matching the `Pods` managed by that `StatefulSet`?

The problem was that the generated YAML contained the status field, not the assertion in kuttl.

I took a look what `json.NewYAMLSerializer#Encode` does behind the scenes, and it actually also runs `json.Marshal` -> `yaml.Unmarshal` -> `yaml.Marshal`. `json.Marshal` is required to respect the json tags in the structs, and `yaml.Unmarshal` and `yaml.Marshal` to convert JSON to YAML. I've implemented the same now in `generate/main.go`, and remove the `status` field before the last step.

iblancasa commented:
> > > Unfortunately the e2e test currently fails on Kubernetes 1.21 because it cannot parse the availableReplicas field in status of v1.StatefulSet.
> > 
> > 
> > So, instead of matching the `StatefulSet`, how about matching the `Pods` managed by that `StatefulSet`?
> 
> The problem was that the generated YAML contained the status field, not the assertion in kuttl.
> 
> I took a look what `json.NewYAMLSerializer#Encode` does behind the scenes, and it actually also runs `json.Marshal` -> `yaml.Unmarshal` -> `yaml.Marshal`. `json.Marshal` is required to respect the json tags in the structs, and `yaml.Unmarshal` and `yaml.Marshal` to convert JSON to YAML. I've implemented the same now in `generate/main.go`, and remove the `status` field before the last step.

Awesome! 🚀
